app/lib/core/Auth/AbstractLDAPAuthAdapter.php

```diff
@@ -0,0 +1,306 @@
+        }
+
+        $va_attrs = ldap_get_attributes($this->getLinkIdentifier(), $vo_entry);
+
+        return array(
+            'user_name' => $ps_username,
+            'email' => $va_attrs[$this->getConfigValue("ldap_attribute_email")][0],
+            'fname' => $va_attrs[$this->getConfigValue("ldap_attribute_fname")][0],
+            'lname' => $va_attrs[$this->getConfigValue("ldap_attribute_lname")][0],
+            'active' => $this->getConfigValue("ldap_users_auto_active"),
+            'roles' => array_merge($this->getConfigValue("ldap_users_default_roles", array()), $this->getRolesToAddFromDirectory($ps_username)),
+            'groups' => array_merge($this->getConfigValue("ldap_users_default_groups", array()), $this->getGroupsToAddFromDirectory($ps_username))
+        );
+    }
+    # --------------------------------------------------------------------------------
+    public function createUserAndGetPassword($ps_username, $ps_password) {
+        // We don't create users in directories, we assume they're already there
+
+        // TODO FIXME The following is insecure!
+        // We will create a password hash that is compatible with the CaUsers authentication adapter though
+        // That way users could, in theory, turn off LDAP authentication later. The hash will not be used
+        // for authentication in this adapter though.
+        return create_hash($ps_password);
+    }
+    # --------------------------------------------------------------------------------
+    protected function getLinkIdentifier() {
+        return $this->opo_ldap;
+    }
+    # --------------------------------------------------------------------------------
+    protected function getConfigValue($ps_key, $pm_default_value = null) {
+        $vm_result = $this->opo_auth_config->get($ps_key);
+        if ($pm_default_value && !$vm_result) {
+            $vm_result = $pm_default_value;
+        }
+        return $vm_result;
+    }
+    # --------------------------------------------------------------------------------
+    protected function getProcessedConfigValue($ps_key, $ps_user_group_name, $ps_user_ou, $ps_base_dn) {
+        $result = $this->getConfigValue($ps_key);
+        $result = str_replace('{username}', $ps_user_group_name, $result);
+        $result = str_replace('{groupname}', $ps_user_group_name, $result);
+        $result = str_replace('{user_ou}', $ps_user_ou, $result);
+        $result = str_replace('{base_dn}', $ps_base_dn, $result);
+        return $result;
+    }
+    # --------------------------------------------------------------------------------
+    public function supports($pn_feature) {
+        switch($pn_feature){
+            case __CA_AUTH_ADAPTER_FEATURE_AUTOCREATE_USERS__:
+                return true;
+            case __CA_AUTH_ADAPTER_FEATURE_RESET_PASSWORDS__:
+            case __CA_AUTH_ADAPTER_FEATURE_UPDATE_PASSWORDS__:
+            default:
+                return false;
```

Comment:
Note this has changed for AD, previously `__CA_AUTH_ADAPTER_FEATURE_RESET_PASSWORDS__` and `__CA_AUTH_ADAPTER_FEATURE_UPDATE_PASSWORDS__` were `true`, but it doesn't make sense to allow a user to change password when they are authenticating with an external Directory.  (OpenLDAP implementation was already returning `false` for both of these).